### PR TITLE
hodge-podge of ideas

### PR DESCRIPTION
**Pull request by trollfot**
Hello Pyrates,

Sorry for the flurry of PR, i had some free time to experiment and here are the results.
This is a hodge-podge of different ideas, with working tests.

In no particular order:
- There is no response set on the protocol: the HTTP request had errors and it failed prior to reaching the app itself, the error is returned immediately. The HTTP request was successful, the app is called and the response is created there. It's then set as the result of the task.
- Added a small cleanup method, used if the protocol needs to stay alive. After the app task callback, the request is reset. The upgrade is kept though, since it was explicitly asked.
- The response can now compute itself.
- The lookup of the route is now a separate method, to allow an easy override (with the idea in mind of maybe making the routing optional and have another mean of resolving the request, like traversing or any kind of graph parkour)
- The extensions now return app, so they can be chained

We can discuss any of it, you're free to dismiss it all too, if nothing is deemed worthy.

### CODE REVIEW

**Comment by davidbgk** [comment]
Hey @trollfot,

You're such a machine! I'm starting to feel a bit ashamed for not being able to keep up with your really pertinent work 😅.

A few thoughts that came after reviewing a previous PR with @yohanboniface but it works for that one too:

* maybe we should try not to be too generic for now and consider that websockets will not be an extension but first-class citizens of Roll. I guess it will reduce a bit the overload associated with the addition (for now).
* I'm surprised all these changes have no impact on performances, I need to test if that's the case with my local configuration and maybe other @pyrates too to validate with different environments, that's not something we used to do but for such a major rewiring it looks pertinent.

That being said, I'll make a review for particular points that caught my attention.

Thanks again! 😎

File: roll/__init__.py [hunk 1]
```diff
@@ -300,91 +345,201 @@ def cookies(self):
...
+
+            response += b'\r\n'
+            if body:
+                response += body
+        else:
+            response += b'\r\n'
+        return response
+
+
+# PROTOCOL STATUS FLAGS
```
Comment (davidbgk) [inline roll/__init__.py]: This probably needs a documentation, at least to understand why we jump from 2 to 4. Reading documentation from https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism I don't get why we have to do that.

File: roll/__init__.py [hunk 2]
```diff
@@ -300,91 +345,201 @@ def cookies(self):
             self._cookies = self.app.Cookies()
         return self._cookies
 
+    def __bytes__(self):
```
Comment (davidbgk) [inline roll/__init__.py]: Tricky but elegant 😁 If I remember correctly we evaluated that option with @yohanboniface but I cannot remember why we discarded it at the end. Did you remember mate?

File: roll/websockets.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+from websockets import handshake, WebSocketCommonProtocol, InvalidHandshake
+from websockets import ConnectionClosed  # exposed for convenience
+
+from roll import ProtocolUpgrade
+
+
+def create_websocket(
+        transport, subprotocol,
+        websocket_timeout=5,
```
Comment (davidbgk) [inline roll/websockets.py]: It looks to be difficult to override these settings without copy/pasting the whole `WebsocketProtocol.__call__` method, maybe we should define directly constants attached to the `WebsocketProtocol` class for easier configuration?

File: roll/websockets.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    )
+    websocket.subprotocol = subprotocol
+    websocket.connection_made(transport)
+    websocket.connection_open()
+    return websocket
+
+
+class WebsocketProtocol(ProtocolUpgrade):
+
+    __slots__ = ('websocket')
```
Comment (davidbgk) [inline roll/websockets.py]: No exposure of `subprotocol`? (dummy question)

File: roll/websockets.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    def connection_lost(self, protocol, exc):
+        self.websocket.connection_lost(exc)
+    connection_lost.bubble_up = True
+
+    def data_received(self, protocol, data):
+        # Received data. We refuse the data if the websocket is
+        # already closed. If the websocket is closing, this data
+        # might be part of the closing handshake (closing frame)
+        if self.websocket.state != 3:  # not closed
```
Comment (davidbgk) [inline roll/websockets.py]: Can we set that magic number as a constant somewhere?

**Comment by trollfot** [comment]
Thank you, @davidbgk 
I'm no machine (I hope ?), just got a bit of time I wished to invest into creative work.
The whole request upgrade process could be more or less generic, depending on the level of abstraction. I don't think the complexity rose too much, with the introduction of the ProtocolUpgrade object, but I'm not the right judge for that.
Regarding the performances, the last commits had a small impact  (~about 10% loss) but I found the request/response flow a bit more natural and less technically-driven, without compromising the ease of testing.

File: roll/websockets.txt [hunk 6]
```diff
@@ -0,0 +1,69 @@
+Websockets inner and outer workings
```
Comment (davidbgk) [inline roll/websockets.txt]: This documentation is valuable, is there a reference page where it came from? Can we set/reuse something alike `http.HTTPStatus` to deal with codes and error messages in a more user-friendly way? (like when we raise errors, I'm pretty sure I'll prefer to receive the extended version of the error 😉)

**Comment by trollfot** [comment]
On my machine, this very branch:

trollfot@valkyrie:~/projects/devel/roll_upgrade$ wrk -t20 -c100 -d30s http://127.0.0.1:3579
Running 30s test @ http://127.0.0.1:3579
  20 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    19.21ms    4.72ms  60.35ms   67.23%
    Req/Sec   261.50     57.28   464.00     60.92%
  156318 requests in 30.03s, 11.63MB read
Requests/sec:   5205.80
Transfer/sec:    396.54KB
trollfot@valkyrie:~/projects/devel/roll_upgrade$

File: tests/test_response.py [hunk 7]
```diff
@@ -46,7 +46,7 @@
     async def get(req, resp):
         resp.status = HTTPStatus.NO_CONTENT
 
-    await client.get('/test')
+    response = await client.get('/test')
```
Comment (davidbgk) [inline tests/test_response.py]: Out of curiosity, does it change anything given that it's not in use?

**Comment by trollfot** [comment]
The current master:

```
trollfot@valkyrie:~/projects/devel/vanilla_roll$ wrk -t20 -c100 -d30s http://127.0.0.1:3579
Running 30s test @ http://127.0.0.1:3579
  20 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    19.00ms    4.20ms  47.82ms   66.93%
    Req/Sec   264.24     50.16   404.00     69.68%
  157962 requests in 30.03s, 11.75MB read
Requests/sec:   5259.94
Transfer/sec:    400.66KB
trollfot@valkyrie:~/projects/devel/vanilla_roll$ 
```

The results are less consistent with my branch, they tend to vary +/- 10%

**Comment by davidbgk** [comment]
> On my machine, this very branch

How does it compare with `master`?

**Comment by davidbgk** [comment]
> The whole request upgrade process could be more or less generic, depending on the level of abstraction. I don't think the complexity rose too much, with the introduction of the ProtocolUpgrade object, but I'm not the right judge for that.

After reviewing it a bit more in depth, I had to confess that it's tempting to pave the way for HTTP2 😇

File: roll/websockets.txt (hunk 6 above)
Comment (trollfot) [reply roll/websockets.txt]: It's from the RFC: https://tools.ietf.org/html/rfc6455
When i was fiddling with the websocket flow, i kep that .txt as a reminder, it should be gone or pushed into a docs/. I kept it here for you, guys, as a quick reference.

File: tests/test_response.py (hunk 7 above)
Comment (trollfot) [reply tests/test_response.py]: That's a left over of many, many, MANY changes i did with the tests and app callback... It needs thorough cleaning, most probably.

File: roll/websockets.py (hunk 5 above)
Comment (trollfot) [reply roll/websockets.py]: My apologies, it exists in the "websockets" package
https://github.com/aaugustin/websockets/blob/master/websockets/protocol.py#L33

File: roll/websockets.txt (hunk 6 above)
Comment (davidbgk) [reply roll/websockets.txt]: Thanks! Too bad it's not a bit more [verbose on the websockets side](https://github.com/aaugustin/websockets/blob/f83362ef91a6ac2ea89fd8bda4d18ff22ffb7a40/websockets/exceptions.py#L118) for easy reuse, that's a “just” a PR-away though.

File: roll/websockets.py (hunk 3 above)
Comment (trollfot) [reply roll/websockets.py]: Yes, i could not yet decide where they should live... The WebsocketProtocol would be the obvious choice.

File: roll/__init__.py (hunk 1 above)
Comment (trollfot) [reply roll/__init__.py]: I'm ashamed to say that it's yet again a left over from when they were used as bit masks...
No reason !

Comment (davidbgk) [reply roll/__init__.py]: No shame, that's why code reviews are so important 😉

**Comment by trollfot** [comment]
The upgrade_delegator now considers a NotImplementedError raised by a surrogate as a bubbling up to the delegating method. We could have considered to make the different methods optional, but I find that an explicit raise is better. A surrogate method that actually does something can still be marked with "bubble_up", to allow the passing through. It wouldn't be "natural" to make it raise some kind of error we'd catch then. But it's still a possibility to harmonize the bubbling process

**Comment by trollfot** [comment]
Back to using a simple return value for the upgrade_delegator bubbling up. It avoid unnecessary checks and is the easiest, syntactically.

Introduced the keep_alive timeout, that produces a 408 Request Timeout error if a given period of time passes without any requests happening on a keep_alive connection. There is not yet a way to configure that. It's still being investigated as it can bring some heaviness into the code. On the other hand, if we are to handle keep_alive, we need it, or we might have zombie connections lying around.

**Comment by trollfot** [comment]
The app now has a set of active connections. It could be useful to cancel remaining watchers on opened protocols, when the server closes. It's being worked on, but as the functionality works, I push it forward to get your feedback regarding its usefulness or not in the project.

**Comment by trollfot** [comment]
The work on this branch mixes new ideas and a refining of the protocol thing. But, if we want to integrate it to the core and focus on actually doing a real swicth of protocol (set_protocol), then if can be started anew

**Pull request closed without merging**